racket/collects/racket/list.rkt

```diff
@@ -588,6 +590,96 @@
     (vector-set! a j x))
   (vector->list a))
 
+(define (combinations l [k #f])
+  (for/list ([x (in-combinations l k)]) x))
+
+;; Generate combinations of the list `l`.
+;; - If `k` is a natural number, generate all combinations of size `k`.
+;; - If `k` is #f, generate all combinations of any size (powerset of `l`).
+(define (in-combinations l [k #f])
+  (unless (list? l)
+    (raise-argument-error 'in-combinations "list?" 0 l))
+  (when (and k (not (exact-nonnegative-integer? k)))
+    (raise-argument-error 'in-combinations "(or #f exact-nonnegative-integer?)" 1 k))
+  (let* ([v       (list->vector l)]
+         [N       (vector-length v)]
+         [N-1     (- N 1)])
+    (letrec ([gen-combinations
+              (cond
+                [(not k)
+                 ;; Enumerate all binary numbers [1..2**N].
+                 ;; Produce the combination with elements in `v` at the same
+                 ;;  positions as the 1's in the binary number.
+                 (let* ([num-incrs (- (expt 2 N) 1)]
+                        [b*        (make-vector N #f)] ;; Binary counter
+                        [b*-incr
+                         ;; Increment the binary number `b*`
+                         (lambda ()
+                           (if (zero? num-incrs)
+                               #f
+                               (begin (set! num-incrs (- num-incrs 1))
+                                      (let loop ([i 0])
+                                        (if (vector-ref b* i)
+                                            (begin (vector-set! b* i #f) (loop (+ i 1)))
+                                            (vector-set! b* i #t))))))]
+                        [b*-num-true
+                         ;; Count the number of #t in the bit vector
+                         (lambda ()
+                           (let loop ([i N-1] [acc 0])
+                             (if (< i 0)
+                                 acc
+                                 (loop (- i 1) (+ acc (if (vector-ref b* i) 1 0))))))]
+                        [b*->combination
+                         ;; Make a list of `v` elements at positions marked `#t` in `b*`
+                         (lambda ()
+                           (let loop ([i N-1] [acc '()])
```

Comment:
Just noticed, reading this email trail that `shuffle` has only one call in
the examples? How about duplicating that call so people see it returns
different results each time?